packages/runtime/src/plugins/useUpstreamRetry.ts

```diff
@@ -0,0 +1,127 @@
+      const subgraphName =
+        executionRequest &&
+        subgraphNameByExecutionRequest.get(executionRequest);
+      const optsForReq = retryOptions({ url, options, context, subgraphName });
+      if (optsForReq) {
+        const {
+          maxRetries,
+          retryDelay = 1000,
+          shouldRetry = (res) => res.status >= 500 || res.status === 429,
+        } = optsForReq;
+        if (maxRetries > 0) {
+          setFetchFn(
+            function (url, options, context, info): MaybePromise<Response> {
+              let retries = maxRetries + 1;
+              let response: Response;
+              function retry(): MaybePromise<Response> {
+                retries--;
+                try {
+                  if (retries < 0) {
+                    return response;
+                  }
+                  const requestTime = Date.now();
+                  return mapMaybePromise(
+                    fetchFn(url, options, context, info),
+                    (currRes) => {
+                      response = currRes;
+                      let retryAfterSeconds: number | undefined;
+                      const retryAfterHeader =
+                        response.headers.get('Retry-After');
+                      if (retryAfterHeader) {
+                        retryAfterSeconds = parseInt(retryAfterHeader);
+                        if (isNaN(retryAfterSeconds)) {
+                          const dateTime = new Date(retryAfterHeader).getTime();
+                          if (!isNaN(dateTime)) {
+                            retryAfterSeconds = dateTime - requestTime;
+                          }
+                        }
+                      }
+                      let currentRetryDelay: number | undefined;
+                      if (retryAfterSeconds) {
+                        currentRetryDelay = retryAfterSeconds * 1000;
+                      } else if (shouldRetry(response)) {
+                        currentRetryDelay = retryDelay;
+                      }
+                      if (currentRetryDelay) {
+                        return new Promise((resolve) => {
+                          const timeout = setTimeout(() => {
+                            timeouts.delete(timeout);
+                            resolve(retry());
+                          }, retryDelay);
+                          timeouts.add(timeout);
+                        });
+                      } else {
+                        return response;
+                      }
+                    },
+                    (e) => {
+                      if (retries < 0) {
+                        throw e;
+                      }
+                      return retry();
+                    },
+                  );
+                } catch (e) {
+                  if (retries < 0) {
+                    throw e;
+                  }
+                  return retry();
+                }
+              }
+              return retry();
+            },
```

Comment:
Why is this duplicated ? You can't just let the error buble up to the catch ?

Reply:
`mapMaybePromise` doesn't handle sync errors